src/snowflake/cli/api/cli_global_context.py

```diff
@@ -14,337 +14,126 @@
+    # nativeapp_definition_v2_to_v1. Consider changing the way
+    # this calculation is provided to commands in order to remove
+    # this logic (then make project_definition a non-cloned @property)
+    override_project_definition: ProjectDefinition | None = None
+
+    _definition_manager: DefinitionManager | None = None
+
+    # which properties invalidate our current DefinitionManager?
+    DEFINITION_MANAGER_DEPENDENCIES = [
+        "project_path_arg",
+        "project_is_optional",
+        "project_env_overrides_args",
+    ]
 
     def reset(self):
         self.__init__()
 
-    @property
-    def enable_tracebacks(self) -> bool:
-        return self._enable_tracebacks
-
-    def set_enable_tracebacks(self, value: bool):
-        self._enable_tracebacks = value
-
-    @property
-    def output_format(self) -> OutputFormat:
-        return self._output_format
-
-    def set_output_format(self, value: OutputFormat):
-        self._output_format = value
-
-    @property
-    def verbose(self) -> bool:
-        return self._verbose
-
-    def set_verbose(self, value: bool):
-        self._verbose = value
-
-    @property
-    def experimental(self) -> bool:
-        return self._experimental
-
-    def set_experimental(self, value: bool):
-        self._experimental = value
-
-    @property
-    def project_definition(self) -> Optional[ProjectDefinition]:
-        return self._project_definition
-
-    def set_project_definition(self, value: ProjectDefinition):
-        self._project_definition = value
+    def clone(self) -> _CliGlobalContextManager:
+        return replace(
+            self,
+            connection_context=self.connection_context.clone(),
+            project_env_overrides_args=self.project_env_overrides_args.copy(),
+        )
 
-    @property
-    def project_root(self):
-        return self._project_root
+    def __setattr__(self, prop, val):
+        if prop in self.DEFINITION_MANAGER_DEPENDENCIES:
+            self._clear_definition_manager()
 
-    def set_project_root(self, project_root: Path):
-        self._project_root = project_root
+        super().__setattr__(prop, val)
 
     @property
-    def project_path_arg(self) -> Optional[str]:
-        return self._project_path_arg
+    def project_definition(self) -> ProjectDefinition | None:
+        if self.override_project_definition:
+            return self.override_project_definition
 
-    def set_project_path_arg(self, project_path_arg: str):
-        self._project_path_arg = project_path_arg
+        return self._definition_manager_or_raise().project_definition
 
     @property
-    def project_env_overrides_args(self) -> dict[str, str]:
-        return self._project_env_overrides_args
-
-    def set_project_env_overrides_args(
-        self, project_env_overrides_args: dict[str, str]
-    ):
-        self._project_env_overrides_args = project_env_overrides_args
+    def project_root(self) -> Path | None:
```

Comment:
this always returns a `Path`
```suggestion
    def project_root(self) -> Path:
```